Resource API refactoring and shared resources

joamaki opened the pull request:
First commit simplifies the Resource[T] API and event type and fixes an issue with coalescing events.
Rest of the commits add k8s.SharedResourceCell with resources for node, services and namespaces and refactor the the relevant pkg/k8s/watchers to use the resources instead of  their own informers and stores.

See individual commits for details.

joamaki committed:
k8s: Introduce shared resources

This introduces a set of centrally managed resource.Resource[T] objects. These aim
to replace the current k8s/watchers and k8s/watchers/subscriber, with a single
interface (Resource[T]) that one can be observed and its store queried from any cell.
With this we'll gradually remove K8sWatcher by moving code from k8s/watchers/ into feature
specific packages that subscribe to Resource[T]. See pkg/k8s/resource/example.go for usage
example of Resource[T].

To start things off, in this patch the Resource[*Service] and Resource[*Node] are introduced
into the object graph and the existing service and node watchers are refactored to be
implemented on top of it. This keeps the existing code working, but allows new cells to
use these objects directly without dependency on K8sWatcher.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Add shared resource for namespaces

Add Resource[*slim_corev1.Namespace] to the shared resources and refactor
the namespace watcher to use it.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on daemon/cmd/cells.go:
```diff
@@ -0,0 +1,63 @@
+
+		// Shared resources provide access to k8s resources as event streams or as
+		// read-only stores.
+		k8s.SharedResourcesCell,
```
Not sure if this should be considered "Infrastructure" or not. It depends on k8s client and our current tests mock that one using fake clients. We could also consider instead mocking `Resource[T]`, but not yet sure if it buys us much. So for now, it's easier to leave this in `ControlPlane`.

joamaki committed:
hive: Fix line wrapping in PrintObjects

The initial implementation was broken as it checked if the line length exceeds width
after printing. Tested manually with long list of objects.

Fixes: 58c6dbf488 ("hive: Detect terminal width for use in InfoLeaf line wrapping")
Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
resource: Refactor the API for simplified usage

As Resource[T] was used more it became apparent that making it an
stream.Observable and having Event[T] be an interface makes it inconvenient
to use:

- Almost always we're subscribing to it for the whole duration of the
  application and do not need to handle completion errors. Consuming the events
  via a channel is also often the preferred way as it always "for-select"
  over multiple resources.

- Event[T] contains almost the same payload for all event types, we might
  as well keep it simple and just use a single struct.

This changes Observe() into `Events(ctx, opts) <-chan Event[T]` and moves
rate limiting and error handling to be per-subscriber options. Event[T]
becomes a single struct with a kind tag to specific event type.

In order to handle large datasets at startup the events are now emitted
immediately rather than waiting for informer to synchronize the store
first.

The BenchmarkResource in resource_test.go was also ported to current master
and the difference between the unbuffered channel approach to callback was
about 2x (2600ns/op for channels, 1350ns/op for callbacks), which is negligble
enough. Benchmark was also ran with buffered event channel with no difference
to unbuffered.  Unbuffered is preferable as allows better chances for coalescing
away the intermediate updates for objects with the same key.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
operator, clustermesh-apiserver: Use k8s.SharedResourcesCell

Use the k8s.SharedResourcesCell in operator and clustermesh-apiserver.
Since resources are lazy, this does not cause unnecessary informers
to run.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
resource: Add finalizer around Event[T].Done to catch misuse

This adds a finalizer to check that the Event[T] has been marked Done()
before it's garbage collected. If Done() is not called we'll end up permanently
blocking events for that key, which can lead to very hard to debug bugs.

The finalizer is set on a heap allocated bool instead of Event[T] itself to
keep that struct as simple as possible with all fields exported, so it can be used in
tests without being exposed to implementation details. The reference chain is:

  Event[T] -> Done() closure -> *bool.

The panic looks like this:

  panic: *resource.resource[*k8s.io/api/core/v1.Node].Events() called from
    /home/jussi/go/src/github.com/cilium/cilium/pkg/k8s/resource/resource_test.go:515 has a
     broken event handler that did not call Done() before event was garbage collected

  goroutine 18 [running]:
  github.com/cilium/cilium/pkg/k8s/resource.(*resource[...]).Events.func1.1()
          /home/jussi/go/src/github.com/cilium/cilium/pkg/k8s/resource/resource.go:343 +0x7d

The BenchmarkResource results with the finalizer is 3900 ns/op versus ~3000 ns/op.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink commented on pkg/k8s/shared_resources.go:
```diff
@@ -0,0 +1,90 @@
+	)
+)
+
+type SharedResources struct {
```
I have to say that this list feels somewhat arbitrary at the moment. What about `v2.CiliumNode` for example which is used in stead of `corev1.Node` depending on the current IPAM mode.

Also why is `LBIPPools` here? Why would we need to share that resource so broadly?

I would say we take a stance, either all resource definitions (that are actually reusable) should be defined here or only the most generic ones. Perhaps we can clarify intended usage with a comment on SharedResources

joamaki replied:
Well, it's not arbitrary, it's just incomplete list. My thinking was that we'd have all of them in one place so it's easy to see to what's all the things we're potentially subscribing to and what all their options are. E.g. it's a bit trying to be a "SharedInformerFactory" in client-go.

There might be cases in the future where we actually don't want to share a resource, e.g. I could imagine that we want `Service` and `Endpoint` to first be coalesced into ServiceCache and that would expose access to them, but for now it feels like it's safer bet to keep all of them in one place.

dylandreimerink commented on pkg/k8s/shared_resources.go:
```diff
@@ -0,0 +1,90 @@
+		return nil, nil
+	}
+	lw := utils.ListerWatcherFromTyped[*corev1.NodeList](cs.CoreV1().Nodes())
+	lw = utils.ListerWatcherWithFields(lw, fields.ParseSelectorOrDie("metadata.name="+nodeTypes.GetName()))
```
The whole local node think doesn't really match the 

> //If you need a filtered view to a resource (e.g. objects in certain namespace) and it already
> // has a more general resource, then use stream.Filter() to filter the event stream rather than
> // creating a new resource.

comment does it? Is there no place where we need info about all of the nodes like in IPCache or something like that?

joamaki replied:
Ah yeah. Plus these don't expose an Observable either. Killing the comment.
No we don't need all `v1.Node`, but we do need all `CiliumNode`s.

dylandreimerink reviewed (changes requested):
Mostly looks alright, a few questions around the actual shared resources cell. Also, would be good to get someone else to take an additional look at the watchers rework in particular since I don't know enough about the existing code.

joamaki committed:
operator/pkg/lbipam: Cleaning up after resource API change

- Add goleak wrapper to the tests
- Add stopping of hive to all the tests to make the goleak check pass.
- Simplify the pre-init/init phases.
- Increase timeouts of few flaky test cases

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
resource: Unset the finalizer as it's more efficient

Instead of checking for the value of the sentinel, it's much
more efficient to just unset the finalizer. Checking the value
was also not correct since there was no memory barrier that would've
prevented reading a stale value.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

nebril reviewed (approved):
Operator changes LGTM.
